examples/run_nlp_trainer.py

```diff
@@ -0,0 +1,350 @@
+
+    return cuda_envs
+
+''' Set up a single GPU CUDA environment on initialization of a mp process pool. '''
+def init_cuda_env(cuda_envs) -> None:
+    try:
+        env=cuda_envs.get()
+        torch.cuda.set_device(env)
+    except (BrokenPipeError, IOError):
+        pass
+
+    # fake a single node world 
+    os.environ['WORLD_SIZE'] = "1"
+    os.environ['LOCAL_WORLD_SIZE'] = "1"
+
+''' List all the checkpoints in a given S3 bucket and folder. ''' 
+def get_all_s3_checkpoints(bucket_name: str, folder_name: str) -> List[str]:
+    # pulls AWS credentials from environment variables, assumes that aws_session_token is defined
+    session = boto3.Session()
+
+    s3 = session.resource('s3')
+    bucket = s3.Bucket(bucket_name)
+    ckpt_names = []
+    for obj in bucket.objects.all():
+        if folder_name in obj.key:
+            obj_name = obj.key
+            ckpt_names.append(obj_name)
+
+    return ckpt_names
+
+''' Consolidate and prettify metrics. '''
+def output_metrics() -> None:
+    tasks = GLUE_METRICS[list(GLUE_METRICS.keys())[0]].keys()
+    large_tasks = ['MNLI', 'QNLI', 'QQP', 'SST-2']
+    # init table headers 
+    headers = ['Checkpoint']
+    headers.extend([f"{task.upper()}" for task in tasks])
+    headers.extend(['GLUE-Large', 'GLUE-All'])
+    tb = [headers]
+
+    # fill table 
+    for ckpt in GLUE_METRICS.keys():
+        output_line = [ckpt]
+        glue_all = 0
+        glue_large = 0
+        # Per task score
+        for task in sorted(GLUE_METRICS[ckpt].keys()):
+            output_line.append("{:.2f}".format(GLUE_METRICS[ckpt][task]))
+            glue_all += GLUE_METRICS[ckpt][task]
+            if task in large_tasks:
+                glue_large += GLUE_METRICS[ckpt][task]
+        # GLUE Large and GLUE All
+        output_line.append("{:.2f}".format(glue_large/len(large_tasks)))
+        output_line.append("{:.2f}".format(glue_all/len(tasks)))
+        tb.append(output_line)
+    
+    print(tabulate(tb, headers='firstrow'))
+
+''' Set up CUDA environment and process pool for given finetuning jobs. '''
+def setup_finetuning_jobs(finetune_tasks: List[str], save_ckpts: List[bool], ckpt_load_path: str, ckpt_filename: str,
+                                         save_folder: str, save_locally: bool, bucket: Optional[str] = None, parent_ckpt: Optional[str] = None, load_ignore_keys: Optional[List[str]] = None) -> None:
+    cuda_envs = init_cuda_queue(len(finetune_tasks))
+    num_tasks = len(finetune_tasks)
+    
+    ''' Callback function for metric collection. '''
+    def log_metrics(metric) -> None:
+        if parent_ckpt: 
+            logged_ckpt_name = parent_ckpt
+        else:
+            logged_ckpt_name = ckpt_filename
+
+        if logged_ckpt_name not in GLUE_METRICS.keys():
+                GLUE_METRICS[logged_ckpt_name] = {}
+        task = list(metric.keys())[0]
+        for m in metric[task]:
+            formatted_task = list(metric.keys())[0].split('_')[1] # remove "glue_" prefix
```

Comment:
I'm wondering whether the `GLUE_METRICS` dictionary is the right data structure -- there seems to be a lot of manual parsing and code that depends on the ordering of the keys. Curious if a `GlueState` object would help make it more readable? Eg.g

```python

@dataclass
class GlueState:
    task_names: List[str]
    task_to_metric: Dict[str, Dict[str, Any]]  # dict of task name to task metric names and values
```

I'm not sure what the right hierarchy is but something like that would help with readability.